File: bindings/python/examples/client_demo.py
#!/usr/bin/env python3
# SPDX-License-Identifier: MIT
# Copyright (c) 2026 huienming

"""Python 客户端示例：流程与 C / C++ / C# 示例一致。

    python examples/client_demo.py [broker] [设备SN] [秒数]
    # 默认：tcp://127.0.0.1:1883、V2023A7B762、6 秒

先起设备端（另开一个窗口）：

    build\\examples\\ncl_device_demo.exe D:\\sim-py 30

probe 模型 → 打印各轴功率/振动含义 → 读 / 写 /STATUS → 订阅采样与事件
→ 按行消费采样（前 8 行）→ 等若干秒 → 打印计数。
"""

import os
import sys
import time

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import nclink  # noqa: E402


def print_axis_quantities(node):
    """把每根轴下的 POWER / ACCELERATION 数据项打出来（含义按"轴 + 物理量"）。"""
    if node.type_name == "AXIS":
        axis = node.name or "轴"
        for item in node.data_items:
            if item.type_name == "POWER":
                print("%-24s %s功率" % (item.path, axis))
            elif item.type_name == "ACCELERATION":
                print("%-24s %s加速度" % (item.path, axis))
    for child in node.devices:
        print_axis_quantities(child)
    for child in node.components:
        print_axis_quantities(child)


def print_sample(sample):
    print("sample %s: id=%s interval=%sms upload=%sms columns=%d rows=%d" % (
        sample.topic, sample.id, sample.interval_ms, sample.upload_interval_ms,
        len(sample.columns), sample.rows))
    for column in sample.columns:
        points_per_slot = column.points // column.slots if column.slots else 0
        print("  %s: %d 个槽位 × 每槽约 %d 点 = %d 点%s" % (
            column.path, column.slots, points_per_slot, column.points,
            "（批量）" if column.is_nested else ""))

    # 按行消费：行数取数据最多的那一列；1 ms 的列在 0.25 ms 的 4 行里读到同一个点。
    shown = min(sample.rows, 8)
    for row in range(shown):
        cells = ["%s=%s" % (column.path, sample.value_at(row, col))
                 for col, column in enumerate(sample.columns)]
        print("  行[%d] %s" % (row, "  ".join(cells)))
    if sample.rows > shown:
        print("  ...（共 %d 行，这里只打前 %d 行）" % (sample.rows, shown))


def print_event(event):
    print("event %s: key=%s value=%s" % (event.topic, event.key, event.value))


def main(argv):
    uri = argv[1] if len(argv) > 1 else "tcp://127.0.0.1:1883"
    sn = argv[2] if len(argv) > 2 else "V2023A7B762"
    seconds = int(argv[3]) if len(argv) > 3 else 6

    nclink.log_init()
    nclink.init(uri)
    print("connected: %s (nclink %s)" % (uri, nclink.version()))

    with nclink.get_device(sn) as device:
        with device.probe() as model:
            print("probe: model root id=%s name=%s" % (model.root.id, model.root.name))
            print("各轴的功率与振动（路径 含义）:")
            for device_node in model.root.devices:
                print_axis_quantities(device_node)

            print("GET /STATUS = %s" % device.get_long("/STATUS"))
            device.set_value("/STATUS", 42)
            print("SET /STATUS = 42 ok")
            print("id(/STATUS) = %s, path(%s) = %s" % (
                device.get_id("/STATUS"), device.get_id("/STATUS"),
                device.get_path(device.get_id("/STATUS"))))

        device.subscribe_samples(2, lambda topic, sample: print_sample(sample))
        device.subscribe_events(2, lambda topic, event: print_event(event))
        print("subscribed: Sample/%s/# and Event/%s" % (sn, sn))

        time.sleep(seconds)
        print("received %d samples, %d events" % (device.sample_count, device.event_count))
        if device.last_callback_error is not None:
            print("callback error: %s" % device.last_callback_error)

    nclink.shutdown()
    nclink.log_shutdown()
    return 0


if __name__ == "__main__":
    try:
        sys.exit(main(sys.argv))
    except nclink.NclinkError as error:
        print("NC-Link error: %s" % error)
        sys.exit(1)

File: examples/client/java/ClientDemo.java
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 huienming

package com.nclink.demo;

import com.nclink.DeviceClient;
import com.nclink.Event;
import com.nclink.Json;
import com.nclink.Model;
import com.nclink.Nclink;
import com.nclink.NclinkException;
import com.nclink.Node;
import com.nclink.Sample;
import com.nclink.SampleColumn;

/**
 * Java 客户端示例：流程与 C / C++ / C# / Python 示例一致。
 *
 * <pre>
 *   java -Djava.library.path=bindings/java/native/bin -cp bindings/java/build/classes \
 *        com.nclink.demo.ClientDemo [broker] [设备SN] [秒数]
 * </pre>
 *
 * 先起设备端（另开一个窗口）：{@code build\examples\ncl_device_demo.exe D:\sim-java 30}
 */
public final class ClientDemo {
    private ClientDemo() {
    }

    public static void main(String[] args) {
        String uri = args.length > 0 ? args[0] : "tcp://127.0.0.1:1883";
        String sn = args.length > 1 ? args[1] : "V2023A7B762";
        int seconds = args.length > 2 ? Integer.parseInt(args[2]) : 6;

        Nclink.logInit();
        try {
            Nclink.init(uri);
            System.out.println("connected: " + uri + " (nclink " + Nclink.version() + ")");

            try (DeviceClient device = Nclink.getDevice(sn)) {
                try (Model model = device.probe()) {
                    System.out.println("probe: model root id=" + model.root().id()
                            + " name=" + model.root().name());
                    System.out.println("各轴的功率与振动（路径 含义）:");
                    for (Node deviceNode : model.root().devices()) {
                        printAxisQuantities(deviceNode);
                    }
                    System.out.println("GET /STATUS = " + device.getLong("/MACHINE/STATUS"));
                    try {
                        device.setValue("/MACHINE/STATUS", "42");
                        System.out.println("SET /STATUS = 42 ok");
                    } catch (NclinkException error) {
                        // 设备没把 /STATUS 绑到 setValue 上就会拒绝，这不是客户端的错
                        System.out.println("SET /STATUS 被拒绝：" + error.name());
                    }
                    String id = device.getId("/MACHINE/STATUS");
                    System.out.println("id(/STATUS) = " + id + ", path(" + id + ") = "
                            + device.getPath(id));

                    try (Json reply = device.methodCall("/plc/getCount", null, true, 5000)) {
                        System.out.println("methodCall(check) = " + reply.encode());
                    }
                }

                device.subscribeSamples(2, (topic, sample) -> printSample(sample));
                device.subscribeEvents(2, (topic, event) -> printEvent(event));
                System.out.println("subscribed: Sample/" + sn + "/# and Event/" + sn);

                for (int i = 0; i < seconds; i++) {
                    Thread.sleep(1000);
                }
                System.out.println("received " + device.sampleCount() + " samples, "
                        + device.eventCount() + " events");
                if (device.lastCallbackError() != null) {
                    System.out.println("callback error: " + device.lastCallbackError());
                }
            }
            Nclink.shutdown();
            Nclink.logShutdown();
        } catch (NclinkException error) {
            System.out.println("NC-Link error: " + error.getMessage());
            System.exit(1);
        } catch (InterruptedException interrupted) {
            Thread.currentThread().interrupt();
        }
    }

    /** 把每根轴下的 POWER / ACCELERATION 数据项打出来（含义按"轴 + 物理量"）。 */
    private static void printAxisQuantities(Node node) {
        if ("AXIS".equals(node.typeName())) {
            String axis = (node.name() == null || node.name().isEmpty()) ? "轴" : node.name();
            for (Node item : node.dataItems()) {
                if ("POWER".equals(item.typeName())) {
                    System.out.printf("%-24s %s功率%n", item.path(), axis);
                } else if ("ACCELERATION".equals(item.typeName())) {
                    System.out.printf("%-24s %s加速度%n", item.path(), axis);
                }
            }
        }
        for (Node child : node.devices()) {
            printAxisQuantities(child);
        }
        for (Node child : node.components()) {
            printAxisQuantities(child);
        }
    }

    private static void printSample(Sample sample) {
        System.out.printf("sample %s: id=%s interval=%sms upload=%sms columns=%d rows=%d%n",
                sample.topic(), sample.id(), sample.intervalMs(),
                sample.uploadIntervalMs(), sample.columns().size(), sample.rows());
        for (SampleColumn column : sample.columns()) {
            System.out.printf("  %s: %d 个槽位 × 每槽约 %d 点 = %d 点%s%n",
                    column.path(), column.slots(),
                    column.slots() > 0 ? column.points() / column.slots() : 0,
                    column.points(), column.isNested() ? "（批量）" : "");
        }

        // 按行消费：行数取数据最多的那一列；1 ms 的列在 0.25 ms 的 4 行里读到同一个点。
        int shown = Math.min(sample.rows(), 8);
        for (int row = 0; row < shown; row++) {
            StringBuilder cells = new StringBuilder();
            for (int col = 0; col < sample.columns().size(); col++) {
                if (col > 0) {
                    cells.append("  ");
                }
                cells.append(sample.columns().get(col).path()).append('=')
                        .append(sample.valueAt(row, col));
            }
            System.out.println("  行[" + row + "] " + cells);
        }
        if (sample.rows() > shown) {
            System.out.println("  ...（共 " + sample.rows() + " 行，这里只打前 " + shown + " 行）");
        }
    }

    private static void printEvent(Event event) {
        System.out.println("event " + event.topic() + ": key=" + event.key()
                + " value=" + event.value());
    }
}
